### PR TITLE
feat(api): collapse spcp/myinfo redirect endpoint to new endpoint

### PR DESCRIPTION
**Pull request by seaerchin**
Part 1 of #1517 

## Problem
Previously, our `/auth/redirect` endpoints were split between the `MyInfo` and `spcp` sections, which made code harder to maintain and our frontend had separated logic. 

Separately, this is also part of the `/api/v3` migration to shift endpoints to their new homes

## Solution
This PR merges the 2 endpoints into 1 and shifts it into a new route rooted at `/api/v3`

**Features**:

1. Created a new controller method under `auth.controller` to handle the redirects for backend
2. Created a new router to service the new `/api/v3` routes
3. New `AuthService` on frontend side as part of react migration 

**Notes**
1. there is no decision making on the frontend now. instead, decisions are deferred to the backend and the frontend just transparently sends across its data
2. `MyInfoService` will be fully deleted in part 2 of this PR. it is left as is first. 

## Tests
Tests are [here](https://github.com/opengovsg/FormSG/pull/1685) to avoid review fatigue

## Manual Tests
1. Create a SP form. Click on activate and go to the form. Click on keep me logged in. Clicking on sign in with singpass should redirect you to the Singpass page. Ensure that after signing in, refreshing the page still keeps you signed in.
2. Create a SP form. Click on activate and go to the form. Clicking on sign in with singpass should redirect you to the Singpass page.
3. Create a email mode form with authentication set to myInfo. Click on activate and go to the form. Clicking on sign in with singpass should redirect to the singpass page.

### CODE REVIEW

File: src/app/modules/auth/auth.controller.ts
Comment (karrui) [inline src/app/modules/auth/auth.controller.ts:243]: I think this function should be in the public-form controller instead? The auth controller is for admin login and logout

File: src/app/modules/auth/auth.controller.ts
Comment (karrui) [inline src/app/modules/auth/auth.controller.ts:315]: I think we should rename this to _handleFormAuthRedirect too, with the underscore making it explicit that it should not be used and is "private".

File: src/app/modules/auth/auth.utils.ts
Comment (karrui) [inline src/app/modules/auth/auth.utils.ts:82]: Same here, this whole function should be in public form utils; and should we also rename it to `mapFormAuthRedirectRouteError`.

Actually, is there a reason why we do not use the usual `mapRouteError`?

File: src/app/modules/spcp/spcp.types.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.types.ts:1]: nit: use relative path

File: src/app/modules/spcp/spcp.types.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.types.ts:49]: nit: rename to `ISpcpForm`, we tend to not use UPPERCASE for abbreviations/initialisms

File: src/public/services/AuthService.ts
Comment (karrui) [inline src/public/services/AuthService.ts:1]: nit: pls call this `PublicFormAuthService`, `AuthService` will be extremely confusing

File: src/public/services/AuthService.ts
Comment (karrui) [inline src/public/services/AuthService.ts:3]: nit: use relative paths

File: src/public/services/AuthService.ts
Comment (karrui) [inline src/public/services/AuthService.ts:6]: I think it's fine to inline them since they are only used once

**Review by karrui — CHANGES_REQUESTED** [review]
some changes requested; pls move the new functions to public-forms module instead of auth
can we also have unit tests at the very least for all the new backend functions?

File: src/app/modules/auth/auth.controller.ts
Comment (karrui) [inline src/app/modules/auth/auth.controller.ts:312]: this is a backwards incompatible change, I'd suggest reverting this.

**Comment by seaerchin** [comment]
@karrui the unit tests are in the linked PR together with integration tests (see tests section) for both FE/BE. This PR is only the logical changes to avoid review fatigue.

File: src/app/modules/auth/auth.controller.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.controller.ts:252]: ohyea, good catch

**Comment by karrui** [comment]
I see that tests are in another PR; fwiw I think unit tests should be contained in the same PR tbh, makes it easier for me

File: src/app/modules/auth/auth.controller.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.controller.ts:243]: ah ok. good point! i'll shift this over to the appropriate folder!

File: src/app/modules/auth/auth.controller.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.controller.ts:312]: sure. i'll still leave the frontend with default parameters but will revert this change to make it backwards compatible

Comment (karrui) [reply src/app/modules/auth/auth.controller.ts:312]: Will need to revert the `!!` change since now there is a possibility for undefinedness

File: src/app/modules/auth/auth.utils.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.utils.ts:82]: i think the main motivating factor for not using the `mapRouteError` was the fear that there might be identical errors that are handled differently. i think this is premature  on my end and i'll combine it with the appropriate `mapRouteError` in public-forms

File: src/app/modules/auth/auth.controller.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.controller.ts:315]: i think this is good convention and will do this going forwards

File: src/app/modules/auth/auth.utils.ts
Comment (mantariksh) [reply src/app/modules/auth/auth.utils.ts:82]: I separated the two mapRouteError functions in the same module because it was easier to check that the correct errors were being covered for each one. not sure if there's any benefit to smooshing all the errors together into a gigantic switch statement even if they're for different endpoints.

File: src/public/services/AuthService.ts
Comment (seaerchin) [reply src/public/services/AuthService.ts:6]: i would prefer not to for `PUBLIC_FORMS_ENDPOINT` because for `validate`, it's going to be reused but i'll fix it for the rest

File: src/app/modules/auth/auth.utils.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.utils.ts]: i do think that if we want `mapRouteError` to just be a black box and let consumers call and get the appropriate error code/message, it should just be 1 huge switch case HAHA

after checking, i don't think there are any errors that we treat differently for redirect, so i combined it into public-form's `mapRouteError`

Comment (mantariksh) [reply src/app/modules/auth/auth.utils.ts]: >if we want mapRouteError to just be a black box and let consumers call and get the appropriate error code/message

I don't understand your point. we do want this - that's what a helper function is for - but why combine the switch cases for unrelated endpoints?

Comment (seaerchin) [reply src/app/modules/auth/auth.utils.ts]: wouldn't the value of `mapRouteError` be for the module where it's rooted in - i.e., that for all routes rooted at `/forms`, we treat all errors similarly and defer the appropriate response to `mapRouteError`?

Comment (mantariksh) [reply src/app/modules/auth/auth.utils.ts]: it's a minor thing so I'm not going to belabour this discussion anymore, but I don't think it's the right approach to anyhow throw unrelated errors together in the same switch case. it makes it so much harder to check that the errors are being covered exhaustively.

File: src/app/modules/spcp/spcp.types.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.types.ts:52]: Suggest a composite type so the original type is preserved:

```ts
export type SpcpForm<T extends IFormSchema> = T & {
  authType: AuthType.SP | AuthType.CP
  esrvcId: string
}
```

File: src/app/modules/spcp/spcp.util.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.util.ts:233]: With the composite type above, if we make this into a type guard instead we can get better typings and remove the need for type casting

```ts
const isSpcpForm = <F extends IFormSchema>(form: F): form is SpcpForm<F> => {
  return [AuthType.SP, AuthType.CP].includes(form.authType)
}
```

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (karrui) [inline src/public/modules/forms/base/directives/submit-form.directive.js:91]: you don't even need the `authType` parameter here right?

**Review by karrui — CHANGES_REQUESTED** [review]
a couple more changes; pls test on staging/local

File: src/app/modules/auth/auth.utils.ts
Comment (karrui) [reply src/app/modules/auth/auth.utils.ts]: @seaerchin i defer to anti, you decide if you want to extract out the mapXError function to a specific form auth one. Fwiw I also agree with him

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (mantariksh) [inline src/app/modules/form/public-form/public-form.controller.ts:415]: I think this business logic (only SP forms can have persistent login) should be in a service

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (mantariksh) [inline src/app/modules/form/public-form/public-form.controller.ts:426]: am I right to assume that providing the generic types is necessary so that `mapErr` and `map` have the correct types?

File: src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts
Comment (mantariksh) [inline src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts:59]: wow how does this not throw an error? myinfogovclient definitely throws if params are missing

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (mantariksh) [inline src/public/modules/forms/base/directives/submit-form.directive.js:115]: I don't think this is a regression because we don't allow logins in preview mode anyway, but could you help double check that we're not losing any functionality here?

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (seaerchin) [reply src/app/modules/form/public-form/public-form.controller.ts:426]: yeap! in general, we actually need to hint when we construct `ok` or `err` directly. this is because they are type constructors and take 2 arguments, 1 of which can never occur.

File: src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts
Comment (seaerchin) [reply src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts:59]: i think this is allowed because we call `jest.mock` on the containing module. this transforms MyInfoGovClient into a mock function, whose return value is then set later on!

Comment (mantariksh) [reply src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts:59]: ah so you're calling `new` on the mocked module. got it

File: src/app/modules/spcp/spcp.errors.ts
Comment (seaerchin) [reply src/app/modules/spcp/spcp.errors.ts:103]: i didn't do this because `MyInfoNoESrvcIdError` has another callsite hwich is not related to this (and is `myInfo` specific). however, if this is desired, we can still keep this error and create a new sum type that is either `{Spcp | MyInfo}...Error`. 

i think you're correct and this serves to make code more readable so i'll go with teh above change!

File: src/app/modules/auth/auth.utils.ts
Comment (seaerchin) [reply src/app/modules/auth/auth.utils.ts]: ok sure! will do

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (seaerchin) [reply src/public/modules/forms/base/directives/submit-form.directive.js:91]: good catch! thanks for reminding me and i'll remove it

File: src/app/modules/spcp/spcp.errors.ts
Comment (mantariksh) [reply src/app/modules/spcp/spcp.errors.ts:103]: you can just use `FormAuthNoEsrvcIdError`, I don't think it's necessary to keep the MyInfo one separate

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (seaerchin) [reply src/app/modules/form/public-form/public-form.controller.ts:415]: good point - i think this should be part of `SpcpFactory.createRedirectUrl` and i'll shift it there!

File: src/app/modules/spcp/spcp.util.ts
Comment (seaerchin) [reply src/app/modules/spcp/spcp.util.ts:233]: good suggestion! will also do teh same for `MyInfo` and look to preserve type information through typeguards in the future

File: src/app/modules/spcp/spcp.errors.ts
Comment (seaerchin) [reply src/app/modules/spcp/spcp.errors.ts:103]: sure, this sounds good and i'll make the changes accordingly! on that note, there's a `MyInfoAuthTypeError` along with `AuthTypeMismatchError` (in spcp) - i think i'll combine these also into `AuthTypeMismatchError` and shift it out into public-forms!

File: src/app/modules/spcp/spcp.errors.ts
Comment (seaerchin) [reply src/app/modules/spcp/spcp.errors.ts:113]: good catch! i'll combine this with the existing `AuthTypeMismatchError`!

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (seaerchin) [reply src/public/modules/forms/base/directives/submit-form.directive.js:115]: yeap, checked staging behaviour with prod - we present users with a view of the logged in form so we aren't losing functionality. users also can't logout (all buttons have no effect besides edit form)

File: src/app/modules/myinfo/myinfo.util.ts
Comment (karrui) [inline src/app/modules/myinfo/myinfo.util.ts:305]: Doesn't seem like it, but ensure that the additional `!!form.esrvcId` check is not a regression!

File: src/app/modules/spcp/spcp.util.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.util.ts:246]: same here

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (karrui) [inline src/public/modules/forms/base/directives/submit-form.directive.js:96]: `MyInfoService` seems to now be unused, should delete in this PR too

File: src/types/api/auth.ts
Comment (karrui) [inline src/types/api/auth.ts:1]: nit: change the name to `PublicFormAuthRedirectDto`?

**Review by karrui — APPROVED** [review]
lgtm with comments

File: src/app/modules/myinfo/myinfo.util.ts
Comment (seaerchin) [reply src/app/modules/myinfo/myinfo.util.ts:305]: good catch! will do so (also for the below one)

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (seaerchin) [reply src/public/modules/forms/base/directives/submit-form.directive.js:96]: `MyInfoService` has `validateESrvcId`, which is why i left it here cos i wanted to delete the whole file at once HAHA more satisfying but i can delete `createRedirectURL` if you want!

File: src/types/api/auth.ts
Comment (seaerchin) [reply src/types/api/auth.ts:1]: np will do

File: src/public/modules/forms/base/directives/submit-form.directive.js
Comment (karrui) [reply src/public/modules/forms/base/directives/submit-form.directive.js:96]: delete the unused code

File: src/app/modules/form/form.errors.ts
Comment (mantariksh) [inline src/app/modules/form/form.errors.ts:78]: nit: to be more precise, you can add that the form has SP, CP or MyInfo enabled but is missing an e-service ID

File: src/app/modules/myinfo/myinfo.util.ts
Comment (seaerchin) [reply src/app/modules/myinfo/myinfo.util.ts:305]: on prod, we don't allow the eserviceId field to be empty. this would get caught by the first layer of checks (and the correct error returned). hence, this isn't a regression!

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (mantariksh) [inline src/app/modules/form/public-form/public-form.controller.ts:447]: should we validate for the form ID in `Segments.PARAMs` as well? I know we already validate using Express route  matching, but it's nice to have all the validation in one place

File: src/app/modules/myinfo/myinfo.util.ts
Comment (mantariksh) [inline src/app/modules/myinfo/myinfo.util.ts:302]: technically a MyInfo form can exist without an e-service ID as long as it's not public. can we name this more specifically to reflect that? eg isMyInfoFormWithEsrvcId

File: src/public/modules/forms/base/directiveViews/submit-form.directive.view.html
Comment (mantariksh) [inline src/public/modules/forms/base/directiveViews/submit-form.directive.view.html:14]: is this set on the `scope`? I don't think it is, which means it is always `undefined`

**Review by mantariksh — CHANGES_REQUESTED** [review]
possible bug with persistent logins. have you tested that persistent logins are working?

File: src/public/modules/forms/base/directiveViews/submit-form.directive.view.html
Comment (mantariksh) [reply src/public/modules/forms/base/directiveViews/submit-form.directive.view.html:14]: ah nvm, I see that it's using `&` binding

**Review by mantariksh — APPROVED** [review]
approving because I don't think the `isPersistentLogin` thing is a bug, but do test that extended logins work before merging

File: src/public/modules/forms/base/directiveViews/submit-form.directive.view.html
Comment (karrui) [reply src/public/modules/forms/base/directiveViews/submit-form.directive.view.html:14]: angularjs strikes again

**Comment by seaerchin** [comment]
@mantariksh could you clarify abit on the extended login? to be safe, i'll push this to staging and test it there!

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (seaerchin) [reply src/app/modules/form/public-form/public-form.controller.ts:447]: oh sure! just to make sure we're on the same wavelength, the express regex should still remain also right?

Comment (mantariksh) [reply src/app/modules/form/public-form/public-form.controller.ts:447]: yep

File: src/app/modules/myinfo/myinfo.util.ts
Comment (seaerchin) [reply src/app/modules/myinfo/myinfo.util.ts:302]: oh yea, definitely. good catch and i'll do so!

**Comment by mantariksh** [comment]
@seaerchin extended login means the Singpass cookie should expire after 30 days instead of the usual 1 hour

File: src/public/modules/forms/base/directiveViews/submit-form.directive.view.html
Comment (seaerchin) [reply src/public/modules/forms/base/directiveViews/submit-form.directive.view.html:14]: the binding thing is still very trippy tbh. just leaving [this](https://www.codelord.net/2016/05/13/understanding-angulars-and-binding/) here for the weird & binding thing in case anyone needs it (i.e, future me)

**Comment by seaerchin** [comment]
@mantariksh gotcha, will do!

**Pull request merged**
